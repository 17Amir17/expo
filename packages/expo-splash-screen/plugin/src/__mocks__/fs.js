// __mocks__/fs.js

const fsMock = jest.createMockFromModule('fs');
fsMock.promises = {};
// Mock fs promise methods
fsMock.promises.readFile = () => Promise.resolve('Mocked content');
fsMock.promises.writeFile = () => Promise.resolve();

module.exports = fsMock;
