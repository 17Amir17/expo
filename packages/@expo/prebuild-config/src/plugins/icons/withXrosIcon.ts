// https://developer.apple.com/documentation/Xcode/configuring-your-app-icon
// VisionAppIcon.solidimagestack
// - Contents.json
// {
//   "info" : {
//     "author" : "expo",
//     "version" : 1
//   },
//   "layers" : [
//     {
//       "filename" : "Front.solidimagestacklayer"
//     },
//     {
//       "filename" : "Middle.solidimagestacklayer"
//     },
//     {
//       "filename" : "Back.solidimagestacklayer"
//     }
//   ]
// }
// - Front.solidimagestacklayer
// -- Contents.json
// {
//   "info" : {
//     "author" : "expo",
//     "version" : 1
//   }
// }
// -- Content.imageset
// --- Contents.json
// {
//   "images" : [
//     {
//       "filename" : "image.png",
//       "idiom" : "reality",
//       "scale" : "2x"
//     }
//   ],
//   "info" : {
//     "author" : "expo",
//     "version" : 1
//   }
// }
// --- image.png
// - Middle.solidimagestacklayer
// - Back.solidimagestacklayer
import { ConfigPlugin, withDangerousMod, withXcodeProject } from '@expo/config-plugins';
import fs from 'fs';
import path from 'path';

import { writeContentsJsonAsync } from './AssetContents';
import { generateXrIconLayerAsync } from './withIosIcons';

// Background must be opaque
// All must be 1024x1024

// Order is especially important!
const XR_ICON_IMAGE_LAYERS = [
  {
    name: 'Front',
  },
  {
    name: 'Middle',
  },
  {
    name: 'Back',
    opaque: true,
  },
];

export const withXrosIcon: ConfigPlugin<{ front: string; middle: string; back: string }> = (
  config,
  props
) => {
  const xcassetName = 'VisionAppIcon';

  withXcodeProject(config, (config) => {
    // TODO: Link to the pbxproj
    return config;
  });

  return withDangerousMod(config, [
    'ios',
    async (config) => {
      await createXrosIconAsync(config.modRequest.projectRoot, {
        rootAssetsDirectory: path.join(
          config.modRequest.projectRoot,
          config.modRequest.projectName!,
          // TODO: Modern Xcode uses `Assets.xcassets`
          'Images.xcassets'
        ),
        xcassetName,
        layers: [props.front, props.middle, props.back],
      });
      return config;
    },
  ]);
};

export async function createXrosIconAsync(
  projectRoot: string,
  options: {
    xcassetName: string;
    rootAssetsDirectory: string;
    /** front, middle, back */
    layers: [string, string, string];
  }
) {
  const imgAssetFolder = path.join(
    options.rootAssetsDirectory,
    options.xcassetName + '.solidimagestack'
  );

  await Promise.all([
    writeContentsJsonAsync(imgAssetFolder, {
      layers: XR_ICON_IMAGE_LAYERS.map((layer) => ({
        filename: layer.name + '.solidimagestacklayer',
      })),
    }),
    ...XR_ICON_IMAGE_LAYERS.map(async (type, index) => {
      const layerAssetDir = path.join(imgAssetFolder, type.name + '.solidimagestacklayer');
      const contentAssetDir = path.join(layerAssetDir, 'Content.imageset');

      const data = await generateXrIconLayerAsync(projectRoot, {
        icon: options.layers[index],
        cacheKey: 'xr-' + type.name,
        removeTransparency: type.opaque ?? false,
      });

      return Promise.all([
        writeContentsJsonAsync(layerAssetDir),
        writeContentsJsonAsync(contentAssetDir, {
          images: data.images,
        }),
        fs.promises.writeFile(path.join(contentAssetDir, data.asset.filename), data.asset.source),
      ]);
    }),
  ]);
}
