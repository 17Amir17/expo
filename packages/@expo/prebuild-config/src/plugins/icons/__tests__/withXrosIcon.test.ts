import * as fs from 'fs';
import { vol } from 'memfs';
import * as path from 'path';

import rnFixture from '../../__tests__/fixtures/react-native-project';
import { createXrosIconAsync } from '../withXrosIcon';

const fsReal = jest.requireActual('fs') as typeof fs;

jest.mock('fs');

describe(createXrosIconAsync, () => {
  const iconPath = path.resolve(__dirname, '../../__tests__/fixtures/icon.png');

  const projectRoot = '/app';
  beforeAll(async () => {
    const icon = fsReal.readFileSync(iconPath);

    vol.fromJSON(rnFixture, projectRoot);

    vol.mkdirpSync('/app/assets');
    vol.writeFileSync('/app/assets/front.png', icon);
    vol.writeFileSync('/app/assets/middle.png', icon);
    vol.writeFileSync('/app/assets/back.png', icon);
  });

  afterAll(() => {
    vol.reset();
  });

  it(`create xros icons`, async () => {
    await createXrosIconAsync(projectRoot, {
      xcassetName: 'VisionAppIcon',
      rootAssetsDirectory: '/app/ios/HelloWorld/Images.xcassets',
      layers: ['/app/assets/front.png', '/app/assets/middle.png', '/app/assets/back.png'],
    });

    const prefix = '/app/ios/HelloWorld/Images.xcassets/VisionAppIcon.solidimagestack/';
    const after = vol.toJSON();
    const icons = Object.fromEntries(
      Object.entries(after)
        .filter(([value]) => value.startsWith(prefix))
        .map(([key, value]) => [key.replace(prefix, ''), value])
    );

    const filenames = Object.keys(icons);
    expect(filenames.length).toEqual(10);
    expect(filenames).toMatchInlineSnapshot(`
      [
        "Front.solidimagestacklayer/Content.imageset/App-Icon-1024x1024@2x.png",
        "Front.solidimagestacklayer/Content.imageset/Contents.json",
        "Front.solidimagestacklayer/Contents.json",
        "Middle.solidimagestacklayer/Content.imageset/App-Icon-1024x1024@2x.png",
        "Middle.solidimagestacklayer/Content.imageset/Contents.json",
        "Middle.solidimagestacklayer/Contents.json",
        "Back.solidimagestacklayer/Content.imageset/App-Icon-1024x1024@2x.png",
        "Back.solidimagestacklayer/Content.imageset/Contents.json",
        "Back.solidimagestacklayer/Contents.json",
        "Contents.json",
      ]
    `);

    const contentsJson = Object.fromEntries(
      Object.entries(icons).filter(([filename]) => filename.endsWith('Contents.json'))
    );

    // Ensure all Contents.json files are valid json with an info object.
    expect(
      Object.values(contentsJson).every((contents) => !!JSON.parse(contents as string).info)
    ).toBe(true);

    expect(sortKeysAscending(contentsJson)).toMatchSnapshot();
  });
});

function sortKeysAscending(dict: Record<string, any>): Record<string, any> {
  return Object.fromEntries(Object.entries(dict).sort(([a], [b]) => a.localeCompare(b)));
}
