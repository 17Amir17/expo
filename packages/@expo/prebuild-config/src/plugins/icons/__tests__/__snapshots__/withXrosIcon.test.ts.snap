// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createXrosIconAsync create xros icons 2`] = `
{
  "Back.solidimagestacklayer/Content.imageset/Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  },
  "images": [
    {
      "filename": "App-Icon-1024x1024@2x.png",
      "idiom": "reality",
      "scale": "2x"
    }
  ]
}",
  "Back.solidimagestacklayer/Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  }
}",
  "Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  },
  "layers": [
    {
      "filename": "Front.solidimagestacklayer"
    },
    {
      "filename": "Middle.solidimagestacklayer"
    },
    {
      "filename": "Back.solidimagestacklayer"
    }
  ]
}",
  "Front.solidimagestacklayer/Content.imageset/Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  },
  "images": [
    {
      "filename": "App-Icon-1024x1024@2x.png",
      "idiom": "reality",
      "scale": "2x"
    }
  ]
}",
  "Front.solidimagestacklayer/Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  }
}",
  "Middle.solidimagestacklayer/Content.imageset/Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  },
  "images": [
    {
      "filename": "App-Icon-1024x1024@2x.png",
      "idiom": "reality",
      "scale": "2x"
    }
  ]
}",
  "Middle.solidimagestacklayer/Contents.json": "{
  "info": {
    "version": 1,
    "author": "expo"
  }
}",
}
`;
